rust/headless-client/src/linux.rs

```diff
@@ -174,37 +187,51 @@ fn parse_resolvectl_output(s: &str) -> Vec<IpAddr> {
         .collect()
 }
 
+async fn run_debug_ipc_client(_cli: Cli) -> Result<()> {
+    tracing::info!(pid = std::process::id(), "run_debug_ipc_client");
+    let stream = UnixStream::connect(SOCK_PATH)
+        .await
+        .context("couldn't connect to UDS")?;
+    let mut stream = IpcStream::new(stream, LengthDelimitedCodec::new());
+
+    stream.send(serde_json::to_string("Hello")?.into()).await?;
+    Ok(())
+}
+
 async fn run_daemon(_cli: Cli) -> Result<()> {
-    let sock_path = dirs::runtime_dir()
-        .context("Failed to get `runtime_dir`")?
-        .join("dev.firezone.client_ipc");
-    ipc_listen(&sock_path).await
+    tracing::info!("run_daemon");
+    ipc_listen().await
 }
 
-async fn ipc_listen(sock_path: &Path) -> Result<()> {
+async fn ipc_listen() -> Result<()> {
+    // Find the `firezone` group
+    let fz_gid = nix::unistd::Group::from_name("firezone")
+        .context("can't get group by name")?
+        .context("firezone group must exist on the system")?
+        .gid;
+
     // Remove the socket if a previous run left it there
-    tokio::fs::remove_file(sock_path).await.ok();
-    let listener = UnixListener::bind(sock_path).unwrap();
+    tokio::fs::remove_file(SOCK_PATH).await.ok();
+    let listener = UnixListener::bind(SOCK_PATH).context("Couldn't bind UDS")?;
+    // UID 0 should be root. The daemon currently must run as root to control DNS.
+    std::os::unix::fs::chown(SOCK_PATH, Some(0), Some(fz_gid.into()))
+        .context("can't set firezone as the group for the UDS")?;
```

Comment:
I don't think we can check the peer's secondary groups easily, and I noticed that Docker uses filesystem permissions to protect its socket, so I think we can also rely on that. I am not sure if Docker tries to do any extra layers of defense.